pkg/compact/downsample/downsample.go

```diff
@@ -0,0 +1,585 @@
+type memBlock struct {
+	// Dummies to implement unused methods.
+	tsdb.IndexReader
+
+	symbols  map[string]struct{}
+	postings []uint64
+	series   []*series
+	chunks   []chunkenc.Chunk
+}
+
+func newMemBlock() *memBlock {
+	return &memBlock{symbols: map[string]struct{}{}}
+}
+
+func (b *memBlock) addSeries(s *series) {
+	sid := uint64(len(b.series))
+	b.postings = append(b.postings, sid)
+	b.series = append(b.series, s)
+
+	for _, l := range s.lset {
+		b.symbols[l.Name] = struct{}{}
+		b.symbols[l.Value] = struct{}{}
+	}
+
+	for i, cm := range s.chunks {
+		cid := uint64(len(b.chunks))
+		s.chunks[i].Ref = cid
+		b.chunks = append(b.chunks, cm.Chunk)
+	}
+}
+
+func (b *memBlock) Postings(name, val string) (index.Postings, error) {
+	allName, allVal := index.AllPostingsKey()
+
+	if name != allName || val != allVal {
+		return nil, errors.New("unsupported call to Postings()")
+	}
+	sort.Slice(b.postings, func(i, j int) bool {
+		return labels.Compare(b.series[b.postings[i]].lset, b.series[b.postings[j]].lset) < 0
+	})
+	return index.NewListPostings(b.postings), nil
+}
+
+func (b *memBlock) Series(id uint64, lset *labels.Labels, chks *[]chunks.Meta) error {
+	if id >= uint64(len(b.series)) {
+		return errors.Wrapf(tsdb.ErrNotFound, "series with ID %d does not exist", id)
+	}
+	s := b.series[id]
+
+	*lset = append((*lset)[:0], s.lset...)
+	*chks = append((*chks)[:0], s.chunks...)
+
+	return nil
+}
+
+func (b *memBlock) Chunk(id uint64) (chunkenc.Chunk, error) {
+	if id >= uint64(len(b.chunks)) {
+		return nil, errors.Wrapf(tsdb.ErrNotFound, "chunk with ID %d does not exist", id)
+	}
+	return b.chunks[id], nil
+}
+
+func (b *memBlock) Symbols() (map[string]struct{}, error) {
+	return b.symbols, nil
+}
+
+func (b *memBlock) SortedPostings(p index.Postings) index.Postings {
+	return p
+}
+
+func (b *memBlock) Index() (tsdb.IndexReader, error) {
+	return b, nil
+}
+
+func (b *memBlock) Chunks() (tsdb.ChunkReader, error) {
+	return b, nil
+}
+
+func (b *memBlock) Tombstones() (tsdb.TombstoneReader, error) {
+	return tsdb.EmptyTombstoneReader(), nil
+}
+
+func (b *memBlock) Close() error {
+	return nil
+}
+
+func downsampleRaw(lset labels.Labels, data []sample, window int64) []*series {
+	if len(data) == 0 {
+		return nil
+	}
+	n := targetChunkSize(len(data))
+
+	countSeries := newAggrSeries(lset, "count", n)
+	sumSeries := newAggrSeries(lset, "sum", n)
+	minSeries := newAggrSeries(lset, "min", n)
+	maxSeries := newAggrSeries(lset, "max", n)
+	counterSeries := newAggrSeries(lset, "counter", 0)
+
+	res := make([]*series, 0, 5)
+
+	if !isCounter(lset) {
+		downsampleSum(sumSeries, data, window)
+		downsampleMin(minSeries, data, window)
+		downsampleMax(maxSeries, data, window)
+		res = append(res, sumSeries, minSeries, maxSeries)
+	}
+	downsampleCount(countSeries, data, window)
+	downsampleCounter(counterSeries, data, window, true)
+	res = append(res, countSeries, counterSeries)
+
+	for _, s := range res {
+		s.close()
```

Comment:
Why we cannot close them in proper `downsampleXXX` function?

Reply:
I generally feel that the scope responsible for creating should also own the closing.

Reply:
ok, let's leave it that way then (: